shared/util/tf/tf_generator.go

```diff
@@ -0,0 +1,439 @@
+package tf
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/chanzuckerberg/happy/shared/config"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/gocty"
+)
+
+type provider struct {
+	Name    string
+	Source  string
+	Version string
+}
+
+type variable struct {
+	Name        string
+	Type        string
+	Description string
+	Default     cty.Value
+}
+
+var requiredProviders []provider = []provider{
+	{
+		Name:    "aws",
+		Source:  "hashicorp/aws",
+		Version: ">= 4.45",
+	},
+	{
+		Name:    "kubernetes",
+		Source:  "hashicorp/kubernetes",
+		Version: ">= 2.16",
+	},
+	{
+		Name:    "datadog",
+		Source:  "datadog/datadog",
+		Version: ">= 3.20.0",
+	},
+	{
+		Name:    "happy",
+		Source:  "chanzuckerberg/happy",
+		Version: ">= 0.53.5",
+	},
+}
+
+var requiredVariables []variable = []variable{
+	{
+		Name:        "aws_account_id",
+		Type:        "string",
+		Description: "AWS account ID to apply changes to",
+	},
+	{
+		Name:        "k8s_cluster_id",
+		Type:        "string",
+		Description: "EKS K8S Cluster ID",
+	},
+	{
+		Name:        "k8s_namespace",
+		Type:        "string",
+		Description: "K8S namespace for this stack",
+	},
+	{
+		Name:        "aws_role",
+		Type:        "string",
+		Description: "Name of the AWS role to assume to apply changes",
+	},
+	{
+		Name:        "image_tag",
+		Type:        "string",
+		Description: "Please provide an image tag",
+	},
+	{
+		Name:        "image_tags",
+		Type:        "string",
+		Description: "Override the default image tags (json-encoded map)",
+		Default:     cty.StringVal("{}"),
+	},
+	{
+		Name:        "stack_name",
+		Type:        "string",
+		Description: "Happy Path stack name",
+	},
+	{
+		Name:        "wait_for_steady_state",
+		Type:        "bool",
+		Description: "Should terraform block until k8s deployment reaches a steady state?",
+		Default:     cty.BoolVal(true),
+	},
+}
+
+const requiredTerraformVersion = ">= 1.3"
+
+type TfGenerator struct {
+	happyConfig *config.HappyConfig
+}
+
+func NewTfGenerator(happyConfig *config.HappyConfig) TfGenerator {
+	return TfGenerator{
+		happyConfig: happyConfig,
+	}
+}
+
+func (tf *TfGenerator) GenerateMain(srcDir, moduleSource string, variables []Variable) error {
```

Comment:
I'll be honest, this function is not my favorite. I would have no idea how to make changes to it or update it and I don't necessarily see the advantages of use an HCL generator over a straight go template. We have this very large switch case anyway, so even though we are doing some dynamic stuff, we'll still need to maintain this switch statement, which is no different then maintaining a template.

You're going to have to convince me on this approach.

Reply:
Can you go over what steps I need to do if I want to upgrade my TF stack version?

And also, what steps would I need to do as a happy developer if we introduced a new stack variable or removed a stack variable? Sounds like we'd need to make changes to the StackConfig struct right?

Reply:
For instance, why not do what we do with fogg apply for module_source, and then add a JSON file with this switch case? I feel like editing a JSON file would be much simpler

Reply:
I can collapse the switch case further, as many of the variables are handled the same way, but not all. There are unique cases:

* References to variables
* Variables with values that come from happy config or module defaults
* Variables with static values
* Special cases (`services`) which are nested of unknown structure

The main goal is to have a source of truth, which template is not. Let's go over it tomorrow.

Reply:
Also, if new variables are added or removed, tf_generator will handle it, as it is both aware of module variables (including their in-depth structure), defaults (for individual object fields as well), and defaults specified in happy config. The only exception are bootstrapping vars, like `app_name` or `stack_name` -- if these variables exist on the module, we're deliberately setting their values.